x/stakeibc/keeper/msg_server_create_trade_route.go

```diff
@@ -2,81 +2,187 @@ package keeper
+		TargetDenomOnTradeZone:  msg.HostDenomOnTrade,
+		TargetDenomOnHostZone:   msg.HostDenomOnHost,
+		HostToRewardHop:         hostToRewardHop,
+		RewardToTradeHop:        rewardToTradeHop,
+		TradeToHostHop:          tradeToHostHop,
 		PoolId:                  msg.PoolId,
 		SpotPrice:               "", // this should only ever be set by ICQ so initialize to blank
-		MinSwapAmount:           msg.MinSwapAmount,
-		MaxSwapAmount:           msg.MaxSwapAmount,
+		MinSwapAmount:           sdkmath.NewIntFromUint64(msg.MinSwapAmount),
+		MaxSwapAmount:           sdkmath.NewIntFromUint64(msg.MaxSwapAmount),
 	}
 
-	k.SetTradeRoute(ctx, newTradeRoute)
+	ms.Keeper.SetTradeRoute(ctx, tradeRoute)
 
-	k.Logger(ctx).Info(fmt.Sprintf("create trade route: %s", msg.String()))
 	return &types.MsgCreateTradeRouteResponse{}, nil
 }
+
+func (k Keeper) RegisterTradeRouteICAAccount(
+	ctx sdk.Context,
+	connectionId string,
+	icaAccountType types.ICAAccountType,
+) (account types.ICAAccount, err error) {
+	// Get the chain ID and counterparty connection-id from the connection ID on Stride
+	chainId, err := k.GetChainIdFromConnectionId(ctx, connectionId)
+	if err != nil {
+		return account, err
+	}
+	connection, found := k.IBCKeeper.ConnectionKeeper.GetConnection(ctx, connectionId)
+	if !found {
+		return account, errorsmod.Wrap(connectiontypes.ErrConnectionNotFound, connectionId)
+	}
+	counterpartyConnectionId := connection.Counterparty.ConnectionId
+
+	// Build the appVersion, owner, and portId needed for registration
+	appVersion := string(icatypes.ModuleCdc.MustMarshalJSON(&icatypes.Metadata{
+		Version:                icatypes.Version,
+		ControllerConnectionId: connectionId,
+		HostConnectionId:       counterpartyConnectionId,
+		Encoding:               icatypes.EncodingProtobuf,
+		TxType:                 icatypes.TxTypeSDKMultiMsg,
+	}))
+	owner := types.FormatICAAccountOwner(chainId, icaAccountType)
+	portID, err := icatypes.NewControllerPortID(owner)
+	if err != nil {
+		return account, err
+	}
+
+	// Create the associate ICAAccount object
+	account = types.ICAAccount{
+		ChainId:      chainId,
+		Type:         icaAccountType,
+		ConnectionId: connectionId,
+	}
+
+	// Check if an ICA account has already been created
+	// (in the event that this trade route was removed and then added back)
+	// If so, there's no need to register a new ICA
+	_, channelFound := k.ICAControllerKeeper.GetOpenActiveChannel(ctx, connectionId, portID)
+	icaAddress, icaFound := k.ICAControllerKeeper.GetInterchainAccountAddress(ctx, connectionId, portID)
+	if channelFound && icaFound {
+		account = types.ICAAccount{
+			ChainId:      chainId,
+			Type:         icaAccountType,
+			ConnectionId: connectionId,
+			Address:      icaAddress,
+		}
+		return account, nil
+	}
+
+	// Otherwise, if there's no account already, register a new one
+	if err := k.ICAControllerKeeper.RegisterInterchainAccount(ctx, connectionId, owner, appVersion); err != nil {
+		return account, err
```

Comment:
So if there is no icaFound yet, we fire off an ICA command to register a new account and because this is async we return the account (types.ICAAccount) with no address specified yet... how does it eventually add the address so that the ICAAccount type is usable?